### PR TITLE
Advanced history features

### PR DESCRIPTION
**Pull request by RezaRahemtola**
# Description

Support of `!?string`, `history nb` and `history` localvar.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the needed labels
- [ ] I have linked this PR to an issue
- [x] I have linked this PR to a milestone
- [ ] I have linked this PR to a project
- [ ] I have tested this code
- [ ] I have added / updated tests (unit / functionals / end-to-end / ...)
- [x] I have updated the README and other relevant documents (guides...)
- [ ] I have added sufficient documentation both in code, as well as in the READMEs

### CODE REVIEW

File: src/command/redirections.c [hunk 1]
```diff
@@ -62,7 +62,7 @@ static char *get_redirect_argument_sum(const char *str, const char *redirect,
     if (target == NULL)
         return (NULL);
     total_len = strlen(redirect) + strlen(target) + 1;
-    sum = malloc(sizeof(char) * (total_len));
+    sum = malloc(sizeof(char) * total_len);
```
Comment (Croos3r) [inline src/command/redirections.c:65]: malloc check missing

File: src/utils/history/substitutions.c [hunk 2]
```diff
@@ -32,18 +34,20 @@ void handle_last_substitution(char **input, list_t *history)
...
 
-    if (cmd == NULL) {
+    if (elem == NULL) {
         fprintf(stderr, "%d: %s\n", nb, NO_EVENT);
         free(*input);
         *input = NULL;
     } else {
-        pattern = malloc(sizeof(char) * (my_nbrlen(nb) + 2));
+        len = snprintf(NULL, 0, "!%d", nb) + 1;
+        pattern = malloc(sizeof(char) * len);
```
Comment (Croos3r) [inline src/utils/history/substitutions.c:48]: cheeeeeeeeeeeeeeeeeeeckkkkkkkkkkkkkkkkkkk

**Review by Croos3r — CHANGES_REQUESTED** [review]
ratio

File: src/command/redirections.c (hunk 1 above)
Comment (EdenComp) [reply src/command/redirections.c:65]: vtff²

**Pull request merged**